scipy/stats/_hypotests.py

```diff
@@ -663,3 +666,352 @@ def somersd(x, y=None):
         raise ValueError("x must be either a 1D or 2D array")
     d, p = _somers_d(table)
     return SomersDResult(d, p, table)
+
+
+def _compute_log_combinations(n):
+    """Compute all log combination of C(n, k)"""
+    gammaln_arr = gammaln(np.arange(n + 1) + 1)
+    return gammaln(n + 1) - gammaln_arr - gammaln_arr[::-1]
+
+
+BarnardExactResult = make_dataclass(
+    "BarnardExactResult", [("statistic", float), ("pvalue", float)]
+)
+
+
+def barnard_exact(table, alternative="two-sided", pooled=True, shgo_n=32):
```

Comment:
I would name this `n` to be future proof. This way we are free to change the optimization algorithm without breaking user code.

Reply:
Updates are in aebd2b88e71c4e47a4455992fc2707dc76b2b2f2. Thanks again @tupui